gitlab/tests/objects/test_mro.py

```diff
@@ -0,0 +1,132 @@
+"""
+Ensure objects defined in gitlab.v4.objects have REST* as last item in class
+definition
+"""
+import inspect
+
+import pytest
+
+import gitlab.v4.objects
+
+# Original notes by John L. Villalovos
+#
+# An example of an incorrect definition:
+#     class ProjectPipeline(RESTObject, RefreshMixin, ObjectDeleteMixin):
+#                           ^^^^^^^^^^ This should be at the end.
+#
+# Correct way would be:
+#     class ProjectPipeline(RefreshMixin, ObjectDeleteMixin, RESTObject):
+#                                       Correctly at the end ^^^^^^^^^^
+#
+#
+# Why this is an issue:
+#
+#   When we do type-checking for gitlab/mixins.py we make RESTObject or
+#   RESTManager the base class for the mixins
+#
+#   Here is how our classes look when type-checking:
+#
+#       class RESTObject(object):
+#           def __init__(self, manager: "RESTManager", attrs: Dict[str, Any]) -> None:
+#               ...
+#
+#       class Mixin(RESTObject):
+#           ...
+#
+#       # Wrong ordering here
+#       class Somev4Object(RESTObject, RefreshMixin):
+#           ...
+#
+#   If we actually ran this in Python we would get the following error:
+#          class Somev4Object(RESTObject, Mixin):
+#     TypeError: Cannot create a consistent method resolution
+#     order (MRO) for bases RESTObject, Mixin
+#
+#   When we are type-checking it fails to understand the class Somev4Object
+#   and thus we can't type check it correctly.
+#
+# Almost all classes in gitlab/v4/objects/*py were already correct before this
+# check was added.
+
+
+def test_show_issue():
+    """Test case to demonstrate the TypeError that occurs"""
+
+    class RESTObject(object):
+        def __init__(self, manager: str, attrs: int) -> None:
+            ...
+
+    class Mixin(RESTObject):
+        ...
+
+    with pytest.raises(TypeError) as exc_info:
+        # Wrong ordering here
+        class Somev4Object(RESTObject, Mixin):
+            ...
+
+    # The error message in the exception should be:
+    #   TypeError: Cannot create a consistent method resolution
+    #   order (MRO) for bases RESTObject, Mixin
+
+    # Make sure the exception string contains "MRO"
+    assert "MRO" in exc_info.exconly()
+
+    # Correctly ordered class, no exception
+    class Somev4Object(Mixin, RESTObject):
+        ...
+
+
+def test_mros():
+    """Ensure objects defined in gitlab.v4.objects have REST* as last item in
+    class definition.
+
+    We do this as we need to ensure the MRO (Method Resolution Order) is
+    correct.
+    """
+
+    debug = False
```

Comment:
nitpick: If we're adding all this, I'd prefer it as a pytest option e.g. in conftest.py instead of a hardcoded `debug` value (so people don't accidentally commit it when changing locally):

```python
def pytest_addoption(parser):
    parser.addoption(
        "--debug-print",
        action="store_true",
        default=False
        help="Print debug info in tests",
    )


@pytest.fixture
def debug_print(request):
    return request.config.getoption("--debug-print")
```

Then pass it to the test with `def test_mros(debug_print):` and use it there.

On the other hand I don't know if this is needed since pytest suppresses all this by default (see below) and you can just control it with capsys flags - in that case I'd just remove the check for debug. Up to you ;)

Reply:
I'll just go ahead and remove this debug code.